app/controller.js

```diff
@@ -100,20 +90,35 @@ export default function(state, emitter) {
       state.archive.addFiles(files, maxSize);
     } catch (e) {
       if (e.message === 'fileTooBig' && maxSize < LIMITS.MAX_FILE_SIZE) {
-        state.modal = signupDialog();
-      } else {
-        state.modal = okDialog(
-          state.translate(e.message, {
-            size: bytes(maxSize),
-            count: LIMITS.MAX_FILES_PER_ARCHIVE
-          })
-        );
+        return emitter.emit('signup-cta', 'size');
       }
+      state.modal = okDialog(
+        state.translate(e.message, {
+          size: bytes(maxSize),
+          count: LIMITS.MAX_FILES_PER_ARCHIVE
+        })
+      );
     }
     render();
   });
 
-  emitter.on('upload', async ({ type, dlimit, password }) => {
+  emitter.on('signup-cta', source => {
```

Comment:
new event to simplify creating the dialog from multiple places and tracking the source